src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultApplicationModelProvider.cs

```diff
@@ -433,25 +433,31 @@ protected virtual ParameterModel CreateParameterModel(ParameterInfo parameterInf
                 throw new ArgumentNullException(nameof(parameterInfo));
             }
 
-            // CoreCLR returns IEnumerable<Attribute> from GetCustomAttributes - the OfType<object>
-            // is needed to so that the result of ToArray() is object
             var attributes = parameterInfo.GetCustomAttributes(inherit: true);
-            var parameterModel = new ParameterModel(parameterInfo, attributes);
+            var parameterModel = new ParameterModel(parameterInfo, attributes)
+            {
+                ParameterName = parameterInfo.Name
+            };
 
-            var bindingInfo = BindingInfo.GetBindingInfo(attributes);
-            if (bindingInfo != null)
+            BindingInfo bindingInfo;
+            if (_modelMetadataProvider is ModelMetadataProvider modelMetadataProviderBase)
+            {
+                var modelMetadata = modelMetadataProviderBase.GetMetadataForParameter(parameterInfo);
+                bindingInfo = BindingInfo.GetBindingInfo(attributes, modelMetadata);
+            }
+            else
             {
-                parameterModel.BindingInfo = bindingInfo;
+                // For backward compatibility, if there's a custom model metadata provider that
+                // only implements the older IModelMetadataProvider interface, construct BindingInfo using attributes.
+                bindingInfo = BindingInfo.GetBindingInfo(attributes);
             }
-            else if (IsFormFileType(parameterInfo.ParameterType))
+
+            if (bindingInfo == null && IsFormFileType(parameterInfo.ParameterType))
             {
-                parameterModel.BindingInfo = new BindingInfo
-                {
-                    BindingSource = BindingSource.FormFile,
-                };
+                bindingInfo = new BindingInfo { BindingSource = BindingSource.FormFile, };
```

Comment:
Could we move this into a model details provider? That seems like a more appropriate place for it.